File: problems/53-MaximumSubarray/cpp/MaximumSubarray.cpp
#include "MaximumSubarray.h"
#include <vector>

using namespace std;

int MaximumSubarray::maxSubArray(vector<int> &nums) {
  if (nums.size() < 1)
    return 0;
  int bestSum = nums[0];
  int prefixSum = nums[0];
  for (size_t i = 1; i < nums.size(); i++) {
    if (prefixSum < 0)
      prefixSum = 0;
    int curSum = prefixSum + nums[i];
    if (curSum > bestSum)
      bestSum = curSum;
    prefixSum = curSum;
  }
  return bestSum;
}

File: problems/53-MaximumSubarray/python/MaximumSubarray.py
#!/bin/python3

from __future__ import annotations
import sys
sys.path.append('../../Utility/python')
from typing import List

class Solution:
  def maxSubArray(self, nums: List[int]) -> int:
    if len(nums) < 1:
      return 0
    bestSum: int = nums[0]
    prefixSum: int = nums[0]
    for i in range(1, len(nums)):
      if prefixSum < 0:
        prefixSum = 0
      if prefixSum + nums[i] > bestSum:
        bestSum = prefixSum + nums[i]
      prefixSum += nums[i]
    return bestSum


if __name__ == '__main__':
  soln = Solution()
  testCases = [
    [ -1, 2, -1, 3, -5 ],
    [ -1, -1, -1 ],
    [  ]
  ]

  for test in testCases:
    print(test, "->", soln.maxSubArray(test))

